### ISSUE TITLE
Favorites are duplicated if the application crashes

### ISSUE THREAD

**Issue by tmancey** [brave/brave-ios#149]
### Steps to Reproduce 
  1. Clear all private data within the Settings area
  2. "http://localhost:6571/about/home/#panel=0" is shown below the address bar
  3. Crash the application (see additional information)
  4. Relaunch "Brave"

**Actual result:** Favorites are duplicated

![screen shot 2018-08-10 at 16 19 48](https://user-images.githubusercontent.com/1188719/43967946-a9f591c6-9cbd-11e8-89fa-effae1690f16.png)

**Expected result:** Favorites should not be duplicated

**Reproduces how often:** Everytime

**Brave Version:** 2.0 (18.08.06.15)

**Device details:** iPad Pro 10.5" (iOS 11.4.1)

### Additional Information 

I simulated a crash in `@objc private func onAddBookmarksFolderButton()` by adding the following line of code `CrashSimulator.forceCrash()`

**Comment by kylehickinson** [comment]
So upon investigation there's a little more to this than just duplication on crashing.

This bug would only occur if crashing on first launch. The reason why this happens is because updating the "is first launch" preference doesn't synchronize to disk before the crash. Therefore on next launch this will add default favourites again because it thinks its a first launch again.

Now, relying on just first launch is not a great solution because not only do you have the preference not synchronizing, it's also possible for CoreData not to save to disk yet. So the flag may say we're past, but we have no default favourites if the app happened to crash just after creating them.

A better solution would be to wait for the `NSManagedObjectContextDidSave` and verify we added these default favourites _and_ they have been saved. So in the event of an unlikely launch crash even if the favourites don't add but the flag does synchronize with `UserDefaults`, it will try to add these again.

We have to keep the combination of flag & CoreData observation because a user can delete all of the favourites if they want.

My question is how deep of a check do we want to make to verify we've added all of the default favourites?

1. Simple count check? We added 5 default favourites and the context saved post-launch with 5 favourites
2. Check each bookmark to verify the default favourites are the exact? (This may not even be necessary since I don't know why we would have other favourites added at launch)

**Comment by kylehickinson** [comment]
Will be fixed by @jhreis as part of another ticket (I believe #161)

**Comment by tmancey** [comment]
Great investigation, I will move to product backlog and block this ticket on #161 so we can retest once #161 is done

**Comment by jhreis** [comment]
Should be fixed per: https://github.com/brave/brave-ios/pull/235
